smugls.1
.\"     Title: smugls
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: May 2008
.\"    Manual: smugls
.\"    Source: smugls
.\"
.TH "SMUGLS" "1" "May 2008" "smugls" "smugls"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
smugls - list files in a smugmug.com gallery
.SH "SYNOPSIS"
.HP 7
\fBsmugls\fR [\fB\-\-email\ email@address\fR] [\fB\-\-password\ password\fR] [\fB\-\-album\ album\fR] [\fB\-\-debug\fR] [\fB\-\-version\fR] [\fB\-\-help\fR]
.SH "DESCRIPTION"
.PP
smugls lists all of the files in a specified smugmug\.com gallery\.
.SH "OPTIONS"
.PP
\fB\-\-email email@address\fR
.RS 4
Specify the email address of your smugmug\.com account\.
.RE
.PP
\fB\-\-password password\fR
.RS 4
Specify the password of your smugmug\.com account\.
.RE
.PP
\fB\-\-debug\fR
.RS 4
Print a whole bunch of debugging messages to stdout\.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print version number\.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\.
.RE
.SH "AUTHOR"
.PP
Written by Greg Kroah\-Hartman
<greg@kroah\.com>\.
.SH "SEE ALSO"
.PP
\fBsmugup\fR(1),
\fBsmugdown\fR(1)


ChangeLog
Summary of changes from v002 to v003
============================================

Andreas Jaeger (7):
      zalloc can be a macro
      Add a spec file
      Allow CFLAGS as argument to make
      Handle fgets return value
      Add config parsing
      Prepare Spec file for version 3
      Fix curl warning

Greg Kroah-Hartman (21):
      added zalloc for zeroed out malloc to make things easier
      save off the original url of an image so we can download it if needed
      smugup:  add --quiet option
      s/which/Which/ when asking for an album name
      add a header to the list of files and handle filenames with no name set
      add file list to album and document struct album
      put files that are in an album in the album instead of the session wide list
      add smugdown.c (for downloading files)
      add smugdown to .gitignore
      smugdown: look at the proper list of files in the album
      Add -O2 to CFLAGS
      free the password and email if we have read it from the config file for smugup and smugdown
      smug_core: fix bug with trailing ;
      smug_core: coding style cleanups to make checkpatch happy
      delete old smugbatch.c file
      smugdown: fix coding style issues found by checkpatch
      smugls: fix coding style issues found by checkpatch
      smugup: fix coding style issues found by checkpatch
      make the buffer really start at 0
      smugdown: start to work on dowloading, don't have the correct url just yet...
      fix up some compiler warnings in smug_core


Summary of changes from v001 to v002
============================================

Andreas Jaeger (2):
      Always set the ssl sanity checks
      Typo fix for smugup and smugls

Greg Kroah-Hartman (7):
      fix problem of not setting the filename for the images uploaded
      increment version number...
      added smugls to .gitignore
      remove compiler warnings from smugls
      build both smugup and smugls
      add ability to read album contents to smug_core.c
      make smugls actually list the files...


Summary of changes from v000 to v001
============================================

Greg KH (2):
      make download show numbers in int not double and use only basename
      remove some of the libraies we need

Greg Kroah-Hartman (45):
      added api key
      added smugup-1.1.sh as the framework to work off of
      .gitignore added
      can read the session id properly
      added simplessl.c example curl code
      clean up code a bit, add other needed urls
      add license and option parsing
      add COPYING
      Makefile rewrite.
      fix compiler warning, hiding the fact that we were not reporting the error properly..
      properly read the buffer sent to us by curl
      now we can log out
      set the proper useragent for the curl session
      don't null out the session_id when reading the logout data
      send the request for the list of albums
      start of filename parsing...
      add smugbatch_version.h to .gitignore
      automatically generate smugbatch_version.h
      switch to the 1.2.0 version of the smugmug api
      we can now read albums
      added --debug
      added md5.c and md5.h from coreutils 6.9 release.
      add md5 to build
      get generation of md5 working properly
      added *.swp to .gitignore
      upload works!
      remove simplessl as it's not needed
      added test image to play with
      add progress function
      properly handle the basename of the uploaded file
      add *.jpg to .gitignore
      make debug global in smugbatch.c
      add smug.h and smug_core.c
      add smug_login and smug_logout
      add album function
      add session_alloc() and session_free()
      added smugup.c
      fix memory problem in smug_core.c
      smugup: make the check for login properly...
      Makefile: switch to smugup instead of smugdump
      add smugup to .gitignore
      add help to smugup
      add ability to choose album to upload to
      add version to help
      add release target to Makefile



smugls.c
/*
 * Copyright (C) 2008 Greg Kroah-Hartman <greg@kroah.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include "list.h"
#include "md5.h"
#include "smugbatch_version.h"
#include "smug.h"


int debug;

static void display_help(void)
{
	fprintf(stdout, "smugls - list files in a smugmug.com album\n");
	fprintf(stdout, "Version: " SMUGBATCH_VERSION "\n");
	fprintf(stdout, "Usage:\n");
	fprintf(stdout, "  smugls [options] files\n");
	fprintf(stdout, "options are:\n");
	fprintf(stdout, "  --email email@address\n");
	fprintf(stdout, "  --password password\n");
	fprintf(stdout, "  --album album\n");
	fprintf(stdout, "  --version\n");
	fprintf(stdout, "  --debug\n");
	fprintf(stdout, "  --help\n");
}

static void display_version(void)
{
	fprintf(stdout, "smugls - version %s\n", SMUGBATCH_VERSION);
}

int main(int argc, char *argv[], char *envp[])
{
	static const struct option options[] = {
		{ "debug", 0, NULL, 'd' },
		{ "email", 1, NULL, 'e' },
		{ "password", 1, NULL, 'p' },
		{ "album", 1, NULL, 'a' },
		{ "version", 0, NULL, 'v' },
		{ "help", 0, NULL, 'h' },
		{ }
	};
	struct session *session;
	struct album *album;
	struct filename *filename;
	char *album_title = NULL;
	int retval;
	int option;
	int found_album;

	session = session_alloc();
	if (!session) {
		fprintf(stderr, "no more memory...\n");
		return -1;
	}

	curl_global_init(CURL_GLOBAL_ALL);
	smug_parse_configfile(session);

	while (1) {
		option = getopt_long_only(argc, argv, "de:p:a:h",
					  options, NULL);
		if (option == -1)
			break;
		switch (option) {
		case 'd':
			debug = 1;
			break;
		case 'e':
			if (session->email)
				free(session->email);
			session->email = strdup(optarg);
			dbg("email = %s\n", session->email);
			break;
		case 'p':
			if (session->password)
				free(session->password);
			session->password = strdup(optarg);
			dbg("password = %s\n", session->password);
			break;
		case 'a':
			album_title = strdup(optarg);
			dbg("album_title = %s\n", album_title);
			break;
		case 'v':
			display_version();
			goto exit;
		case 'h':
			display_help();
			goto exit;
		default:
			display_help();
			goto exit;
		}
	}

	if (!session->email) {
		fprintf(stdout, "Enter smugmug.com email address: ");
		session->email = get_string_from_stdin();
	}

	if (!session->password) {
		fprintf(stdout, "Enter smugmug.com password: ");
		session->password = get_string_from_stdin();
	}

	retval = smug_login(session);
	if (retval) {
		fprintf(stderr, "Can not login\n");
		return -1;
	}

	dbg("1\n");
	retval = smug_get_albums(session);
	if (retval) {
		fprintf(stderr, "Can not read albums\n");
		return -1;
	}

	if (!album_title) {
		fprintf(stdout, "Available albums:\n");
		list_for_each_entry(album, &session->albums, entry)
			fprintf(stdout, "\t%s\n", album->title);
		fprintf(stdout, "\nWhich album id to list? ");
		album_title = get_string_from_stdin();
	}

	found_album = 0;
	list_for_each_entry(album, &session->albums, entry) {
		if (strcmp(album->title, album_title) == 0) {
			found_album = 1;
			break;
		}
	}
	if (!found_album) {
		fprintf(stdout, "Album %s is not found\n", album_title);
		return -1;
	}

	retval = smug_read_images(session, album);
	if (retval) {
		fprintf(stderr, "Error uploading files\n");
		return -1;
	}

	/* print out the files */
	fprintf(stdout, "ID        Key   Filename\n");
	list_for_each_entry(filename, &album->files, entry) {
		char *name = filename->filename;
		if (strlen(name) == 0)
			name = "<no name>";
		fprintf(stdout, "%s %s %s\n", filename->id, filename->key,
			name);
	}

	free(album_title);
	smug_logout(session);

	session_free(session);
exit:
	return 0;
}
